features/works/work_download.feature

```diff
@@ -202,3 +202,66 @@ Feature: Download a work
   When I am logged in as a "policy_and_abuse" admin
     And I hide the work "TOS Violation"
   Then I should not see "Download"
+
+  Scenario: Downloads of related work expire when parent work's anonymity changes.
+
+  Given a hidden collection "Hidden"
+    And I have related works setup
+    And I post a related work as remixer
+    And I post a translation as translator
+    And I log out
+  When I view the work "Followup"
+    And I follow "HTML"
+  Then I should see "Worldbuilding by inspiration"
+  When I view the work "Worldbuilding Translated"
+    And I follow "HTML"
+  Then I should see "Worldbuilding by inspiration"
+  # Going from revealed to unrevealed
+  When I am logged in as "inspiration"
+    And I edit the work "Worldbuilding" to be in the collection "Hidden"
+    And I log out
+    And I view the work "Followup"
+    And I follow "HTML"
+  Then I should not see "inspiration"
+    And I should see "Inspired by a work in an unrevealed collection."
+  When I view the work "Worldbuilding Translated"
+    And I follow "HTML"
+  Then I should not see "inspiration"
+    And I should see "Translation of a work in an unrevealed collection."
+  # Going from unrevealed to revealed
+  When I reveal works for "Hidden"
+    And I log out
+    And I view the work "Followup"
+    And I follow "HTML"
+  Then I should see "Worldbuilding by inspiration"
+  When I view the work "Worldbuilding Translated"
+    And I follow "HTML"
+  Then I should see "Worldbuilding by inspiration"
+
+  Scenario: Downloads of related work expire when child work's anonymity changes.
+
+  Given a hidden collection "Hidden"
+    And I have related works setup
+    And a translation has been posted and approved
+    And a related work has been posted and approved
+  When I view the work "Worldbuilding"
+  Then I should see the related work listed on the original work
+    And I should see the translation listed on the original work
+  # Going from revealed to unrevealed
+  When I am logged in as "translator"
+    And I edit the work "Worldbuilding Translated" to be in the collection "Hidden"
+    And I am logged in as "remixer"
+    And I edit the work "Followup" to be in the collection "Hidden"
+    And I view the work "Worldbuilding"
+    And I follow "HTML"
+  Then I should not see the related work listed on the original work
+    And I should not see the translation listed on the original work
+    And I should see "A work in an unrevealed collection."
+  # Going from unrevealed to revealed
+  When I reveal works for "Hidden"
+    And I log out
+    And I view the work "Worldbuilding"
+    And I follow "HTML"
+  Then I should see the related work listed on the original work
```

Comment:
@tickinginstant I'm not sure how to expire the caches of works such that downloads are up to date. I tried poking the work model but the cache of the work appears to be already updating appropriately but not the download.

Reply:
The caching issues aren't reproducible in tests because the caching is nginx-side, rather than Rails-side. The cache expiration is handled by making each download URL append a timestamp:
https://github.com/otwcode/otwarchive/blob/538664f7b1de73539a5c61b28c37c72ca7a0821d/app/helpers/works_helper.rb#L121-L124
So that when the work's `updated_at` is changed, the link to the download changes. nginx's cache keys are based on the path, so when `updated_at` changes, the cache key used by nginx also changes. (It's the same principle behind the way that Rails does non-recyclable cache keys — if the cache key includes the timestamp, then changes to the timestamp won't delete the old cache key, but they will cause the cache key to change, forcing the cache to be regenerated.)

Reply:
Thank you!